### ISSUE TITLE
Crop feature

### ISSUE THREAD

**Issue by FunkyMuse** [panpf/zoomimage#53]
I think that this awesome library can benefit a lot with a crop functionality.

Especially when you zoom in/out, whatever is shown on the screen of the user can be cropped to the bounds of the screen, making it perfect for documents etc...

**Comment by panpf** [comment]
I don't understand your requirement, please describe it in more detail or provide an example for reference

**Comment by FunkyMuse** [comment]
> I don't understand your requirement, please describe it in more detail or provide an example for reference

An example project
https://github.com/alexvasilkov/GestureViews?tab=readme-ov-file

For example as you move/pan/zoom you can just crop the image to the bounds of the level you're panning/zooming

**Comment by panpf** [comment]
I see what you mean, it should be easy to add a crop() function that returns a Bitmap containing only the pixels in the currently visible area. Is this a big deal? Must it be included in the version that is going to be released this week?

**Comment by FunkyMuse** [comment]
> I see what you mean, it should be easy to add a crop() function that returns a Bitmap containing only the pixels in the currently visible area. Is this a big deal? Must it be included in the version that is going to be released this week?

Yeah, adding a crop() that returns a bitmap containing the pixels currently in the visible area is what the other library does.

It's a big deal when it comes to a library like the one you've built I think, but it's not my place to say the importance of this, it's your time and effort, your decision.

**Comment by panpf** [comment]
I thought about it again and it seems not that easy. There are two situations:

1. When subsampling is supported, you can directly decode region to get the visible pixels

2. When subsampling is not supported, it is more troublesome and I haven’t thought of a simple solution yet

So it may take some time

**Comment by panpf** [comment]
The screenshot function has been launched in the demo. Please pull the code of the dev branch and run the demo to experience it.

The implementation idea is to take a screenshot of the ZoomAsyncImage component and then capture its valid part according to zoomable.contentDisplayRect. Therefore, this function does not require any changes to the zoomimage library. You only need to refer to the implementation in the demo and port it to your App.

Reference code:
* https://github.com/panpf/zoomimage/blob/dev/sample/src/commonMain/kotlin/com/github/panpf/zoomimage/sample/ui/examples/BaseZoomImageSample.kt#L499
* https://github.com/panpf/zoomimage/blob/dev/sample/src/commonMain/kotlin/com/github/panpf/zoomimage/sample/ui/examples/SketchZoomAsyncImageSample.kt#L51
* https://github.com/panpf/zoomimage/blob/dev/sample/src/commonMain/kotlin/com/github/panpf/zoomimage/sample/ui/util/capturable.kt
* https://github.com/panpf/zoomimage/blob/dev/sample/src/nonAndroidMain/kotlin/com/github/panpf/zoomimage/sample/ui/util/uis.nonAndroid.kt#L25

![Snipaste_2025-01-16_17-22-52](https://github.com/user-attachments/assets/a4e9586e-2f16-4c68-8ff0-9a12065e3ae6) ![Snipaste_2025-01-16_17-23-11](https://github.com/user-attachments/assets/0a52957c-827d-41f1-995c-2ea01aced6ac)

**Comment by FunkyMuse** [comment]
> The screenshot function has been launched in the demo. Please pull the code of the dev branch and run the demo to experience it.
> 
> The implementation idea is to take a screenshot of the ZoomAsyncImage component and then capture its valid part according to zoomable.contentDisplayRect. Therefore, this function does not require any changes to the zoomimage library. You only need to refer to the implementation in the demo and port it to your App.
> 
> Reference code:
> 
>     * https://github.com/panpf/zoomimage/blob/dev/sample/src/commonMain/kotlin/com/github/panpf/zoomimage/sample/ui/examples/BaseZoomImageSample.kt#L499
> 
>     * https://github.com/panpf/zoomimage/blob/dev/sample/src/commonMain/kotlin/com/github/panpf/zoomimage/sample/ui/examples/SketchZoomAsyncImageSample.kt#L51
> 
>     * https://github.com/panpf/zoomimage/blob/dev/sample/src/commonMain/kotlin/com/github/panpf/zoomimage/sample/ui/util/capturable.kt
> 
>     * https://github.com/panpf/zoomimage/blob/dev/sample/src/nonAndroidMain/kotlin/com/github/panpf/zoomimage/sample/ui/util/uis.nonAndroid.kt#L25
> 
> 
> ![Snipaste_2025-01-16_17-22-52](https://github.com/user-attachments/assets/a4e9586e-2f16-4c68-8ff0-9a12065e3ae6) ![Snipaste_2025-01-16_17-23-11](https://private-user-images.githubusercontent.com/3250512/403790071-0a52957c-827d-41f1-995c-2ea01aced6ac.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc0OTE2MTIsIm5iZiI6MTczNzQ5MTMxMiwicGF0aCI6Ii8zMjUwNTEyLzQwMzc5MDA3MS0wYTUyOTU3Yy04MjdkLTQxZjEtOTk1Yy0yZWEwMWFjZWQ2YWMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMjFUMjAyODMyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjQ5Y2ViM2ExMjQzZTA2OTRiZDcwNzI5ZTM3OWU5MmRlYjk1MDVkZGFlNjJjZDZhN2E1YWM3NDc5N2VlZTYzZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.UxWHAheOzbt6kcUOudP-yXFX6e6hJaPQo8zq4u4iR10)

trying to check out the branch but keep getting this error

`Could not find io.github.panpf.sketch4:sketch-animated-gif:4.0.0-beta01-SNAPSHOT`

**Comment by panpf** [comment]
The version of sketch4 in dev branch is 4.0.1. Have you not updated to the latest version?

**Comment by panpf** [comment]
Cropping has been added to the web version of the example：https://panpf.github.io/zoomimage/app/

**Comment by FunkyMuse** [comment]
Checked it out in the sample app and in the web version, looks good IMO, great job!

**Comment by FunkyMuse** [comment]
> Good, just port the code from the example to your app!

is this feature scheduled for 1.1.2?

**Comment by panpf** [comment]
There is no need to include this functionality in ZoomImage and it will not be added to ZoomImage in the future. 

My implementation only relies on the contentDisplayRect property of the ZoomImage, unlike [GestureViews](https://github.com/alexvasilkov/GestureViews?tab=readme-ov-file). 

You add the code from the example to your App and look at the code and you'll see what I mean!
